filament/backend/src/vulkan/caching/VulkanDescriptorSet.cpp

```diff
@@ -0,0 +1,646 @@
+    5, 5, 6, 5, 6, 6, 7, 3, 4, 4, 5, 4, 5, 5, 6, 4, 5, 5, 6, 5, 6, 6, 7, 4, 5, 5, 6, 5, 6, 6, 7, 5,
+    6, 6, 7, 6, 7, 7, 8};
+
+inline uint8_t countBits(uint32_t num) {
+    return BIT_COUNT[num & 0xFF] + BIT_COUNT[(num >> 8) & 0xFF] + BIT_COUNT[(num >> 16) & 0xFF] +
+           BIT_COUNT[(num >> 24) & 0xFF];
+}
+
+struct SamplerSet {
+    using Mask = UniformBufferBitmask;
+    struct Key {
+        Mask mask;
+        static_assert(sizeof(Mask) == 4);
+        uint32_t padding;// We need a padding to ensure 8-byte alignement.
+        VkDescriptorSetLayout layout;
+        VkSampler sampler[SAMPLER_BINDING_COUNT];
+        VkImageView imageView[SAMPLER_BINDING_COUNT];
+        VkImageLayout imageLayout[SAMPLER_BINDING_COUNT];
+    };
+
+    struct Equal {
+        bool operator()(Key const& k1, Key const& k2) const {
+            if (k1.mask != k2.mask) return false;
+            if (k1.layout != k2.layout) return false;
+
+            for (uint8_t i = 0, bitCount = countBits(k1.mask); i < bitCount; ++i) {
+                if (k1.sampler[i] != k2.sampler[i] || k1.imageView[i] != k2.imageView[i] ||
+                        k1.imageLayout[i] != k2.imageLayout[i]) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    };
+    using HashFn = utils::hash::MurmurHashFn<Key>;
+    using Cache = std::unordered_map<Key, VulkanDescriptorSet*, HashFn, Equal>;
+};
+
+struct UBOSet {
+    using Mask = SamplerBitmask;
+
+    struct Key {
+        Mask mask;
+        static_assert(sizeof(Mask) == 4);
+        uint32_t padding;// We need a padding to ensure 8-byte alignement.
+        VkDescriptorSetLayout layout;
+        VkBuffer buffers[UNIFORM_BINDING_COUNT];    //   80     0
+        VkDeviceSize offsets[UNIFORM_BINDING_COUNT];//   40  1592
+        VkDeviceSize sizes[UNIFORM_BINDING_COUNT];  //   40  1632
+    };
+
+    struct Equal {
+        bool operator()(Key const& k1, Key const& k2) const {
+            if (k1.mask != k2.mask) return false;
+            if (k1.layout != k2.layout) return false;
+
+            for (uint8_t i = 0, bitCount = countBits(k1.mask); i < bitCount; ++i) {
+                if (k1.buffers[i] != k2.buffers[i] || k1.offsets[i] != k2.offsets[i] ||
+                        k1.sizes[i] != k2.sizes[i]) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    };
+    using HashFn = utils::hash::MurmurHashFn<Key>;
+    using Cache = std::unordered_map<Key, VulkanDescriptorSet*, HashFn, Equal>;
+};
+
+constexpr uint8_t MAX_BINDINGS = UNIFORM_BINDING_COUNT * MAX_SUPPORTED_SHADER_STAGE;
+static_assert(MAX_BINDINGS >= SAMPLER_BINDING_COUNT * MAX_SUPPORTED_SHADER_STAGE);
+
+template<typename MaskType>
+inline MaskType genMask(uint8_t stage, MaskType bit) {
+    return static_cast<MaskType>(bit << (sizeof(MaskType) * 4 * (stage - 1)));
+}
```

Comment:
This logic is a bit confusing to me. Maybe document what the `4` represents and what a "mask" is?

Reply:
This function is actually no longer necessary.  removed.